### PR TITLE
Bug fix for empty inputs on RhythmTransfor. Test Cases not 100% covered

### PR DESCRIPTION
**Pull request by cvf-bcn-gituser**
I have isolated the files that need to be changed related to the bug in RT. (RhythmTransform).
Only Test Cases related to the change are added (TestEmpty,TestZero, etc)

RegressionTest needs to be added and possibly some other use cases, so putting PR in draft mode.

### CODE REVIEW

**Review by dbogdanov — CHANGES_REQUESTED** [review]
I've proposed some changes to the wrong memory access fix, and also to keep in control the correct shape of the input vector.

File: src/algorithms/rhythm/rhythmtransform.cpp [hunk 1]
```diff
@@ -51,53 +51,61 @@ void RhythmTransform::compute() {
...
-  // derive and transpose
-  vector<vector<Real> > bandsDerivative(nBands);
-  for (int band=0; band<nBands; ++band) {
-    vector<Real> derivTemp(nFrames);
-    derivTemp[0] = 0.0;
-    for (int frame=1; frame<nFrames; frame++) {
-      derivTemp[frame] = bands[frame][band]-bands[frame-1][band];
+  // Check first that we have a valid populated input
+  // Otherwise core dump could happen.
+  if ((nFrames!=0) && (nFrames!=0))
```
[the remark this answers is not in the document]
Comment (cvf-bcn-gituser) [reply src/algorithms/rhythm/rhythmtransform.cpp:56]: Yes I agree.I  prefer an early return or "throw"  than having to indent the code  with an extra "if" .
In some coding conventions its forbidden to have return statements anywhere but at the end, and to have it only once.

**Comment by cvf-bcn-gituser** [comment]
Todays commit on rhythmtransform.cpp file to take care of diverse/empty  vector band lengths threw the expected error in the UTs.
On the Linux subsystem the coredump still occurred, I need to check the test cases simulatenously on both platforms tomorrow.

File: src/algorithms/rhythm/rhythmtransform.cpp [hunk 2]
```diff
@@ -49,63 +49,83 @@ void RhythmTransform::compute() {
   vector<vector<Real> >& output = _rhythmTransform.get();
 
   int nFrames = bands.size();
-  int nBands = bands[0].size();
+  // 
+  // Check first that we have a valid populated input to avoid  core dump
+  if (nFrames== 0) {
+    throw EssentiaException("File looks like a completely silent file");
```
Comment (dbogdanov) [inline src/algorithms/rhythm/rhythmtransform.cpp:55]: Use more informative messages for exceptions: 
- Mention the algorithm name (`RhythmTransform: Exception message`) similar other algorithms.
- RhythmTransform does not work with files, therefore this message is misleading. We can say "Input mel-spectrogram is empty".

File: src/algorithms/rhythm/rhythmtransform.cpp [hunk 3]
```diff
@@ -49,63 +49,83 @@ void RhythmTransform::compute() {
...
-    // derive and transpose
-    vector<vector<Real> > bandsDerivative(nBands);
-    for (int band=0; band<nBands; ++band) {
-      vector<Real> derivTemp(nFrames);
-      derivTemp[0] = 0.0;
-      for (int frame=1; frame<nFrames; frame++) {
-        derivTemp[frame] = bands[frame][band]-bands[frame-1][band];
-      }
-      bandsDerivative[band] = derivTemp;
+    std::cout << "A Violation found empty band"<< std::endl;
```
Comment (dbogdanov) [inline src/algorithms/rhythm/rhythmtransform.cpp]: Remove the debug message.

File: src/algorithms/rhythm/rhythmtransform.cpp [hunk 4]
```diff
@@ -49,63 +49,83 @@ void RhythmTransform::compute() {
...
+    throw EssentiaException("File looks like an empty band");
```
Comment (dbogdanov) [inline src/algorithms/rhythm/rhythmtransform.cpp]: Same here: RhythmTransform  doesn't work with files.

File: src/algorithms/rhythm/rhythmtransform.cpp [hunk 5]
```diff
@@ -49,63 +49,83 @@ void RhythmTransform::compute() {
...
+    return;
+  }
+
+  // Check for an inconsistent input vector with inner vectors of different length
+  if(!( std::equal(bandSizes.begin() + 1, bandSizes.end(), bandSizes.begin()) ))
+  {
+    std::cout << "A Violation found different length" << std::endl;
```
Comment (dbogdanov) [inline src/algorithms/rhythm/rhythmtransform.cpp]: Remove debug message

File: src/algorithms/rhythm/rhythmtransform.cpp [hunk 6]
```diff
@@ -49,63 +49,83 @@ void RhythmTransform::compute() {
...
+    throw EssentiaException("Inconsistent input vector with inner vectors of different length");
```
Comment (dbogdanov) [inline src/algorithms/rhythm/rhythmtransform.cpp]: Add algorithm name.

**Review by dbogdanov — CHANGES_REQUESTED** [review]

File: src/algorithms/rhythm/rhythmtransform.cpp (hunk 3 above)
Comment (cvf-bcn-gituser) [reply src/algorithms/rhythm/rhythmtransform.cpp]: This was already removed (not outdated label)

File: src/algorithms/rhythm/rhythmtransform.cpp (hunk 5 above)
Comment (cvf-bcn-gituser) [reply src/algorithms/rhythm/rhythmtransform.cpp]: This was already removed (not outdated label)

File: src/algorithms/rhythm/rhythmtransform.cpp (hunk 6 above)
Comment (cvf-bcn-gituser) [reply src/algorithms/rhythm/rhythmtransform.cpp]: RhythmTransform: label added to all Exceptions thrown

File: test/src/unittests/rhythm/test_rhythmtransform.py [hunk 7]
```diff
@@ -0,0 +1,97 @@
...
+        rt = RhythmTransform()(bands)
+        self.assertEqual(len(rt), 1)
+
+        bands = [ones(512), ones(512), ones(512)]
+        rt = RhythmTransform()(bands)
+        self.assertEqual(len(rt), 1)
+
+
+
+    def testEmpty(self):
```
[the remark this answers is not in the document]
Comment (cvf-bcn-gituser) [reply test/src/unittests/rhythm/test_rhythmtransform.py]: Made some updates and a commit.

Its an "end of day at office commit" to allow continue work on different machine.

File: src/algorithms/rhythm/rhythmtransform.cpp [hunk 8]
```diff
@@ -51,53 +51,61 @@ void RhythmTransform::compute() {
   int nFrames = bands.size();
   int nBands = bands[0].size();
```
[the remark this answers is not in the document]
Comment (cvf-bcn-gituser) [reply src/algorithms/rhythm/rhythmtransform.cpp]: review line 52

**Review by cvf-bcn-gituser — COMMENTED** [review]
Line 45,46  remove the rms prefix, and choose better name.

File: test/src/unittests/rhythm/test_rhythmtransform.py [hunk 9]
```diff
@@ -51,27 +49,42 @@ def testRegression(self):
...
+        # This code stores reference values in a file for later loading.
+        # np.save('rhythmtransform.npy', rt)
+        # Reference samples are loaded as expected values
+        expected_rhythmtransform_npy = np.load(join(filedir(), 'rhythmtransform/rhythmtransform.npy'))
+        expected_rhythmtransform = expected_rhythmtransform_npy.tolist()
 
-        #rt_mean = numpy.mean(rt, axis=0)
-        #bin_resoluion = 5.007721656976744
-        #print("Estimated BPM: %0.1f" % float(numpy.argmax(rt_mean) * bin_resoluion))
+        self.assertAlmostEqualVectorFixedPrecision(expected_rhythmtransform[0], rt[0] ,2)
```
Comment (cvf-bcn-gituser) [inline test/src/unittests/rhythm/test_rhythmtransform.py:72]: Test full matrix and check precisions, push further the precision.

File: test/src/unittests/rhythm/test_rhythmtransform.py [hunk 10]
```diff
@@ -36,9 +36,7 @@ class TestRhythmTransform(TestCase):
     def testInvalidParam(self):
         self.assertConfigureFails(RhythmTransform(), {'frameSize': -1})
         self.assertConfigureFails(RhythmTransform(), {'hopSize': -1})
-    """
-    FIXME: Work in progress for Regression Test
-    """
+
     def testRegression(self):
         # Simple regression test, comparing normal behaviour
         audio = MonoLoader(filename = join(testdata.audio_dir, 'recorded', 'techno_loop.wav'),sampleRate = 44100)()
```
Comment (dbogdanov) [inline test/src/unittests/rhythm/test_rhythmtransform.py:42]: Let's use a better variable naming:
`frameSize` ->  `melFrameSize`
`rmsFrameSize` -> `rtFrameSize`

Rename hop sizes accordingly.

**Review by dbogdanov — CHANGES_REQUESTED** [review]

File: test/src/unittests/rhythm/test_rhythmtransform.py [hunk 11]
```diff
@@ -51,27 +49,42 @@ def testRegression(self):
...
-        assertEqual(len(rt), 1) only checks for whether a certain number of 
-        rhythm-domain frames is output for the given input. 
-        Can we test the expected value of the RD-frame output instead? 
-        What should it be?
-        """    
+    def testNonUniformEmptyInput(self):
+        nEmptySlots = [[],[0,0]]
+        self.assertRaises(EssentiaException, lambda: RhythmTransform()(nEmptySlots))
 
     def testZeros1Darray(self):
```
Comment (cvf-bcn-gituser) [inline test/src/unittests/rhythm/test_rhythmtransform.py:89]: Make names semanitc ZeroMelFrames

File: test/src/unittests/rhythm/test_rhythmtransform.py [hunk 12]
```diff
@@ -80,6 +93,7 @@ def testZeros1Darray(self):
         self.assertEqualVector(listZeros,rt_list)
 
     def testZeros2Darray(self):
+        # non-zero number of input frames, but each mel-frame vector is empty
```
Comment (cvf-bcn-gituser) [inline test/src/unittests/rhythm/test_rhythmtransform.py:96]: rename to testZero2MelFrames

File: test/src/unittests/rhythm/test_rhythmtransform.py (hunk 9 above)
Comment (cvf-bcn-gituser) [reply test/src/unittests/rhythm/test_rhythmtransform.py:72]: done

File: test/src/unittests/rhythm/test_rhythmtransform.py (hunk 10 above)
Comment (cvf-bcn-gituser) [reply test/src/unittests/rhythm/test_rhythmtransform.py:42]: done

File: test/src/unittests/rhythm/test_rhythmtransform.py (hunk 11 above)
Comment (cvf-bcn-gituser) [reply test/src/unittests/rhythm/test_rhythmtransform.py:89]: done

**Pull request merged**